File: frontend/src/game/upgrades/stat_modifiers/bullet_pierce_1.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const BulletPierce1Def: UpgradeDef = {
  id: "bullet_pierce_1",
  name: "Piercing Shot",
  description: "Bullets pierce +1 enemy",
  rarity: RarityID.Rare,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 10,
  targetClass: UpgradeTargetID.Bullet,
  fieldInTargetClass: UpgradeStatID.Pierce,
  value: 1,
  specificAttackType: "bullet",
  stackable: true,
  maxStacks: 2,
}

export class BulletPierce1 extends Upgrade {}

File: backend/app/core/upgrades/stat_modifiers/bullet_pierce_1.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class BulletPierce1(UpgradeImplementation):
    id = 'bullet_pierce_1'
    name = 'Piercing Shot'
    description = 'Bullets pierce +1 enemy'
    rarity = 'rare'
    type = 'stat_modifier'
    cost = 10

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 1
        appliers.modifiers.add_multiplier('bullet', 'pierce', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True
